Implement CRUD Operation Endpoints for User, Customer, Consumer, RerservationTransaction, Reservation, Payment Entities

spatiszisis opened the pull request:
- Removed the flyway implementation. We will introduce it in a future issue.
- Introduced a basic CRUD system for the entities (User, Customer, Consumer, ReservationTransaction, Reservation, and Payment).
- Added an exception handler.

spatiszisis committed:
#3 - Implement CRUD Operation Endpoints for User, Customer, Consumer, RerservationTransaction, Reservation, Payment Entities
- Removed the flyway implementation. We will introduce it in future issue
- Introduced the User entity with all the related CRUD implementations
- Added an exception handler

spatiszisis committed:
#3 - Implement CRUD Operation Endpoints for User, Customer, Consumer, RerservationTransaction, Reservation, Payment Entities
- Introduced a basic CRUD system for the entities (User, Customer, Consumer, ReservationTransaction, Reservation, and Payment)
- Removed the flyway functionality, we will introduce it in a future issue

spatiszisis committed:
#3 - Implement CRUD Operation Endpoints for User, Customer, Consumer, RerservationTransaction, Reservation, Payment Entities
- Added swagger annotations in the controllers

apokiou reviewed (approved):
Looks Good !

EmmanSkout reviewed (approved):
Very good , a minor (maybe) point is that we can possible have a base abstract controller in the future

christosavramis commented on src/main/java/net/trapezokomos/dashboard/service/IAbstractService.java:
```diff
@@ -0,0 +1,11 @@
+import net.trapezokomos.dashboard.exception.ResourceAlreadyExistsException;
+import org.springframework.data.domain.Pageable;
+
+public interface IAbstractService<T, ID> {
```
remove T, replace ID with Entity

christosavramis commented on src/main/java/net/trapezokomos/dashboard/data/BaseDocument.java:
```diff
@@ -0,0 +1,47 @@
+
+@Data
+@MappedSuperclass
+public class BaseDocument {
```
Rename to AbstractEntity

christosavramis commented on src/main/java/net/trapezokomos/dashboard/data/User.java:
```diff
@@ -0,0 +1,73 @@
+)
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
```
@Data may cause performance issues,
use
@Setter @GETTER

christosavramis commented on src/main/java/net/trapezokomos/dashboard/exception/ResourceAlreadyExistsException.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.io.Serial;
+
+public class ResourceAlreadyExistsException extends Exception {
```
Lets create an intermediate Exception Class 

GenericAPIException
GenericAPIRunTimeException
and extend through those

christosavramis commented on db_schema.puml:
```diff
@@ -0,0 +1,133 @@
+@startuml
```
why we need this?

spatiszisis replied:
This is a structure annotation for **.puml** files to know the plugin that reads these files that this is the start of the specific file

christosavramis commented on src/main/java/net/trapezokomos/dashboard/utils/UserConverter.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.Set;
+
+@Component
+public class UserConverter implements AbstractConverter<UserResource, User> {
```
We could also try spring converters interface Converter

christosavramis commented on src/main/java/net/trapezokomos/dashboard/controller/UserController.java:
```diff
@@ -0,0 +1,85 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/user")
```
lets move the api to "api/v1/user" to be able to easily serve html on root or not and have api versioning

spatiszisis committed:
#3 - Implement CRUD Operation Endpoints for User, Customer, Consumer, RerservationTransaction, Reservation, Payment Entities
- Renamed BaseDocument to AbstractEntity
- Changed the import of @Data to @Getter and @Setter
- Added to implement the AttributeConverter interface in order to clone the proper methods
- Added a API version to all the routes
